cloudmarker/alerts/emailalert.py

```diff
@@ -68,10 +79,38 @@ def done(self):
             self.body = ''.join(self.stringbuffer)
         message.attach(MIMEText(self.body))
 
-        smtp_connection = smtplib.SMTP(host=self.host, port=self.port)
+        smtp = self._get_connection()
         try:
-            smtp_connection.sendmail(self.sender, self.to, message.as_string())
+            smtp.sendmail(self.sender, self.to, message.as_string())
         except smtplib.SMTPException as e:
             _log.error('Failed to send email: %s', e)
         finally:
-            smtp_connection.quit()
+            smtp.quit()
+
+    def _get_connection(self):
+        """Return SMTP connection object.
+
+        Create a SMTP connection based on whether the SSL param.
+        If SSL connection is required (for eg: in case of gmail) then
+        method will return a SMTP_SSL connection. In other cases if the
+        authentication is not required or SMTP_SSL connection is not
+        required then a plain SMTP connection object is returned.
+        # For gmail follow the steps given below even to work for 2 factor
+        authentication
+        # 1. Log-in into Gmail with your account
+        # 2. Navigate to https://security.google.com/settings/security/
+        apppasswords
+        # 3. In 'select app' choose 'custom', give it an arbitrary name and
+        press generate
+        # 4. It will give you 16 chars token.
+        # 5. Use that token as password for login
+        # 6. Host is smtp.gmail.com and port is 465
+        """
+        # In case no authentication is required then no need to login
```

Comment:
The `use_ssl` flag should not decide whether authentication should occur or not. Using SSL or not and performing authentication or not are orthogonal concerns. They are independent of each other.

Whether to perform authentication or not should depend on `username` is defined or not. It means, if `username` is defined, then `smtp.login()` should occur in both cases: `use_ssl = True` as well as `use_ssl = False`.

Also, since credentials should never be sent in plaintext, for `use_ssl = False`, we need to invoke `smtp.starttls()` before invoking `smtp.login()`.

Reply:
`SMTP_SSL` connection initiates a TLS encrypted connection. In case we are using a simple `SMTP` connection then in that case we need to encrypt it via `starttls`.

Reference: https://realpython.com/python-send-email/